qa: Improvements to debug_assert_log + busy_wait_for_debug_log

hodlinator opened the pull request:
* Remove incorrect docstring in `busy_wait_for_debug_log()`.
* Replace nerfed regex searches with `X in Y` expressions.
* Only compute the log string to be printed on failure *when we actually fail* instead of every 0.05s.
* As we find each needle (expected message) in the haystack (log output), stop searching for it. **If we fail and time out, we will only complain about the needles (expected messages) we didn't find. On master we also include found needles, which is less helpful.**

Found while developing a new test case in https://github.com/bitcoin/bitcoin/pull/33336#discussion_r2351892330

l0rinc commented on test/functional/test_framework/test_node.py:
```diff
@@ -586,18 +590,21 @@ def busy_wait_for_debug_log(self, expected_msgs, timeout=60):
+                    found.append(i)
 
-            if found:
+            if len(found) == len(expected_msgs):
```
Wouldn't this return if the same expected message appears multiple times in the log? I think we should collect the indexes into a set instead.

l0rinc reviewed (changes requested):
Thanks for reducing useless work - did you measure any speedup?
I think we could make the result simpler and more pythonic - it's a test, I find visibly simpler code better here than theoretically faster one that we can't even measure, so I suggested a few changes.
Here's the end-result containing most of my suggestions:
```patch
diff --git a/test/functional/test_framework/test_node.py b/test/functional/test_framework/test_node.py
index e28140c17b..ccbde4702c 100755
--- a/test/functional/test_framework/test_node.py
+++ b/test/functional/test_framework/test_node.py
@@ -550,32 +550,31 @@ class TestNode():
         time_end = time.time() + timeout * self.timeout_factor
         prev_size = self.debug_log_size(encoding="utf-8")  # Must use same encoding that is used to read() below
 
-        def print_log(log):
-            return " - " + "\n - ".join(log.splitlines())
-
         yield
 
+        def join_lines(logs):
+            return " - " + "\n - ".join(logs.splitlines())
+
+        remaining = list(expected_msgs)
         while True:
-            found = []
             with open(self.debug_log_path, encoding="utf-8", errors="replace") as dl:
                 dl.seek(prev_size)
                 log = dl.read()
+
             for unexpected_msg in unexpected_msgs:
-                if re.search(re.escape(unexpected_msg), log, flags=re.MULTILINE):
-                    self._raise_assertion_error(f'Unexpected message "{unexpected_msg}" found in log:\n\n{print_log(log)}\n\n')
-            for i, expected_msg in enumerate(expected_msgs):
-                if re.search(re.escape(expected_msg), log, flags=re.MULTILINE):
-                    found.append(i)
+                if unexpected_msg in log:
+                    self._raise_assertion_error(f'Unexpected message "{unexpected_msg}" found in log:\n\n{join_lines(log)}\n\n')
+
+            found = {i for i, m in enumerate(remaining) if m in log}
             # We found remaining expected messages before encountering unexpected messages.
-            if len(found) == len(expected_msgs):
+            if len(found) == len(remaining):
                 return
             # Remove the ones we found so far, so we only complain about remaining ones when timing out.
-            for f in reversed(found):
-                expected_msgs.pop(f)
+            remaining = [m for m in remaining if m not in found]
             if time.time() >= time_end:
                 break
             time.sleep(0.05)
-        self._raise_assertion_error(f'Expected message(s) {expected_msgs!s} not found in log:\n\n{print_log(log)}\n\n')
+        self._raise_assertion_error(f'Expected message(s) {remaining!r} not found in log:\n\n{join_lines(log)}\n\n')
 
     @contextlib.contextmanager
     def busy_wait_for_debug_log(self, expected_msgs, timeout=60):
@@ -589,21 +588,16 @@ class TestNode():
 
         yield
 
+        remaining = list(expected_msgs)
         while True:
-            found = []
             with open(self.debug_log_path, "rb") as dl:
                 dl.seek(prev_size)
                 log = dl.read()
 
-            for i, expected_msg in enumerate(expected_msgs):
-                if expected_msg in log:
-                    found.append(i)
-
-            if len(found) == len(expected_msgs):
+            found = {i for i, m in enumerate(remaining) if m in log}
+            if len(found) == len(remaining):
                 return
-
-            for f in reversed(found):
-                expected_msgs.pop(f)
+            remaining = [m for m in remaining if m not in found]
 
             if time.time() >= time_end:
                 print_log = " - " + "\n - ".join(log.decode("utf8", errors="replace").splitlines())
@@ -612,8 +606,7 @@ class TestNode():
             # No sleep here because we want to detect the message fragment as fast as
             # possible.
 
-        self._raise_assertion_error(
-            f'Expected message(s) {expected_msgs!s} not found in log:\n\n{print_log}\n\n')
+        self._raise_assertion_error(f'Expected message(s) {remaining!r} not found in log:\n\n{print_log}\n\n')
 
     @contextlib.contextmanager
     def wait_for_new_peer(self, timeout=5):
```

hodlinator committed:
refactor(qa): Avoid unnecessary string operations

print_log was recalculated every 0.05s in assert_debug_log(), even during successful circumstances - changed to only be computed upon failure.

Simplified terminology from "(does not) partially match(es)" to "(not) found in" since it seems to reference the first function having used regular expression matching, while it always escaped the search strings (see parent commit). (Simplified grammar also avoids issues with singular/plural "was/were not found").

hodlinator committed:
qa: Replace always-escaped regexps with "X in Y"

Always escaping the search string makes the use of regular expressions unnecessary.

Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

hodlinator reviewed (commented):
Thanks for the feedback!

> Thanks for reducing useless work - did you measure any speedup?

Rebased new push on top of #33336 which has a lot of `assert_debug_log()` (010cf81407c0df8de766fd2a116463d180f25f33), the below represents somewhat average runs:
```
₿ git co 2025/09/assert_debug_log_rebased~3
₿ hyperfine -r 30 -N ./build/test/functional/feature_assumevalid.py
Benchmark 1: ./build/test/functional/feature_assumevalid.py
  Time (mean ± σ):      6.058 s ±  0.049 s    [User: 4.263 s, System: 0.520 s]
  Range (min … max):    5.985 s …  6.157 s    30 runs

₿ git co 2025/09/assert_debug_log_rebased
₿ hyperfine -r 30 -N ./build/test/functional/feature_assumevalid.py
Benchmark 1: ./build/test/functional/feature_assumevalid.py
  Time (mean ± σ):      5.998 s ±  0.045 s    [User: 4.237 s, System: 0.496 s]
  Range (min … max):    5.929 s …  6.109 s    30 runs
```
So the tendency is a difference of 1% for this test case - a bit underwhelming. Would expect to see more of a difference for longer log files and more `expected_msgs` per invocation.

hodlinator commented:
CI failure (https://github.com/bitcoin/bitcoin/actions/runs/17885097855/job/50857723211?pr=33423) seems unrelated:

`ERROR: failed to build: failed to solve: short read: expected 659078263 bytes but got 90259897: unexpected EOF`
`Command '['./ci/test/02_run_container.sh']' returned non-zero exit status 1.`

hodlinator commented:
Latest push has somewhat simpler code (no `found`-list). Tested performance again, similarly to https://github.com/bitcoin/bitcoin/pull/33423#pullrequestreview-3244741286:
```
₿ git co 2025/09/assert_debug_log_rebased~3
₿ hyperfine -r 30 -N ./build/test/functional/feature_assumevalid.py
Benchmark 1: ./build/test/functional/feature_assumevalid.py
  Time (mean ± σ):      6.034 s ±  0.047 s    [User: 4.260 s, System: 0.488 s]
  Range (min … max):    5.964 s …  6.146 s    30 runs
 
₿ git co 2025/09/assert_debug_log_rebased
₿ hyperfine -r 30 -N ./build/test/functional/feature_assumevalid.py
Benchmark 1: ./build/test/functional/feature_assumevalid.py
  Time (mean ± σ):      5.990 s ±  0.046 s    [User: 4.233 s, System: 0.482 s]
  Range (min … max):    5.881 s …  6.126 s    30 runs
```
So less than 1% performance improvement in this case. Again, should perform better for other cases.

l0rinc commented on test/functional/test_framework/test_node.py:
```diff
@@ -546,30 +546,36 @@ def assert_debug_log(self, expected_msgs, unexpected_msgs=None, timeout=2):
+                if unexpected_msg in log:
+                    self._raise_assertion_error(f'Unexpected message "{unexpected_msg}" '
+                                                f'found in log:\n\n{join_log(log)}\n\n')
+            while len(remaining_expected) > 0:
+                if remaining_expected[-1] in log:
+                    remaining_expected.pop()
+                else:
+                    break
```
this basically pops until the last element is still in logs, stops otherwise, right?
```suggestion
        while remaining_expected and remaining_expected[-1] in log:
            remaining_expected.pop()
```

l0rinc commented on test/functional/test_framework/test_node.py:
```diff
@@ -546,30 +546,36 @@ def assert_debug_log(self, expected_msgs, unexpected_msgs=None, timeout=2):
+                    remaining_expected.pop()
+                else:
+                    break
+            if len(remaining_expected) == 0:
```
it would be slightly more pythonic to do

```suggestion
            if not remaining_expected:
```

l0rinc commented on test/functional/test_framework/test_node.py:
```diff
@@ -546,30 +546,36 @@ def assert_debug_log(self, expected_msgs, unexpected_msgs=None, timeout=2):
                 break
             time.sleep(0.05)
-        self._raise_assertion_error('Expected messages "{}" does not partially match log:\n\n{}\n\n'.format(str(expected_msgs), print_log))
+        self._raise_assertion_error(f'Expected message(s) {remaining_expected!s} '
+                                    f'not found in log:\n\n{join_log(log)}\n\n')
```
that's not necessarily true, we have just stopped searching after the first failure, right?

hodlinator replied:
Thanks for noticing this issue!

Latest push inserts
```python
remaining_expected = [e for e in remaining_expected if e not in log]
```
before the error reporting for it to be correct, meaning we only need to do the full search on failure.

l0rinc reviewed (commented):
The error message seems off to me now, we're potentially logging lines as missing that we haven't checked.

l0rinc commented on test/functional/test_framework/test_node.py:
```diff
@@ -546,30 +546,36 @@ def assert_debug_log(self, expected_msgs, unexpected_msgs=None, timeout=2):
+                                    f'not found in log:\n\n{join_log(log)}\n\n')
 
     @contextlib.contextmanager
     def busy_wait_for_debug_log(self, expected_msgs, timeout=60):
```
>         Return:
            the number of log lines we encountered when matching

This doesn't seem to be the case anymore

hodlinator replied:
Thanks! Added commit removing this from docstring with reference in commit message to ref commit where it was made incorrect.

hodlinator committed:
qa: Only complain about expected messages that were not found

Only leaves messages we never found in the final assert message of the functions, which is more helpful (remaining_expected).

Avoids repeatedly searching for messages we have already found (pop()).

Stops searching for other expected messages if we already failed finding one. Still need to clean remaining_expected at the end, but *only if we fail*.

Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

hodlinator committed:
doc: Remove no longer correct comment

Left over since fa7b07571f24b6def6effdd4cc1b96c7507bf959.

Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

hodlinator commented on test/functional/test_framework/test_node.py:
```diff
@@ -546,30 +546,36 @@ def assert_debug_log(self, expected_msgs, unexpected_msgs=None, timeout=2):
-                    found = False
-            if found:
+                if unexpected_msg in log:
+                    self._raise_assertion_error(f'Unexpected message "{unexpected_msg}" '
+                                                f'found in log:\n\n{join_log(log)}\n\n')
```
Regarding nit: I sympathize with the way `busy_wait_for_debug_log()` on master puts the parameter to `_raise_assertion_error()` on it's own line, so I would be tempted to do that for all of them - which makes me realize: if only I broke the message in two lines, I could raise up the first half and still keep close to 80 char width - and we're back with the current approach. Left unchanged in latest push.

hodlinator reviewed (commented):
Latest push fixes bug from previous push (https://github.com/bitcoin/bitcoin/pull/33423#discussion_r2379926157), simplifies code (https://github.com/bitcoin/bitcoin/pull/33423#discussion_r2379920746) and adds commit to fix docstring (https://github.com/bitcoin/bitcoin/pull/33423#discussion_r2379958513).

Compared run-times for all functional tests and wasn't able to find one which clearly benefited from this optimization. So I'm open to just replacing `while remaining_expected and remaining_expected[-1] in log: remaining_expected.pop()` with `remaining_expected = [e for e in remaining_expected if e not in log]` if reviewers prefer.

maflcko reviewed (commented):
review ACK a1f76230209629c5141984aaac1cc4ba7b4f761a 💨

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK a1f76230209629c5141984aaac1cc4ba7b4f761a 💨
8T3ac9W58hqzIhMuuFTWcQhtsU1LZVPoRlAGhSUhD6onZEgb+z4MOUskJrtOEwPr8vGl/+uFTacNyZixlVF0Bw==
```

</details>

l0rinc commented:
rfm?
